### ISSUE TITLE
[import/no-unused-modules] False positive while importing TS interfaces

### ISSUE THREAD

**Issue by maneetgoyal** [benmosher/eslint-plugin-import#1924]
1. The use of `import/no-unused-modules` flags raises a false error: **exported declaration 'XYZ' not used within other modules** when using `import type { XYZ } from "file.ts"` notation. 🐛 

2. Source Code:

```ts
// file1.ts

export interface Foo {

}
```

```ts
// file2.ts

import type { Foo } from "./file1.ts";
```

3. Current Behavior: Lint error

```
exported declaration 'Foo' not used within other modules eslint(import/no-unused-modules)
```

4. Expected behavior: No error

---

Other notes:

I could use `import { Foo } from "./file1.ts"` but this coding style conflicts with the [@typescript-eslint/consistent-type-imports](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/consistent-type-imports.md) rule and I get a lint error:

```
Import "NumberFormatProps" is only used as types eslint(@typescript-eslint/consistent-type-imports)
```

Will be really great if `import/no-unused-modules` can be upgraded to handle `import type ...` syntax. 🎉 

---

Attn. https://github.com/benmosher/eslint-plugin-import/pull/1819
cc @nicolashenry @ljharb

**Comment by ljharb** [comment]
`import type` should be ignored, and `export` of a type-space value should imo also be ignored.

**Comment by ljharb** [comment]
I agree, `import type` should be considered here.

**Comment by nicolashenry** [comment]
I think it is probably working under some conditions depending on loading order like Flow type imports but I still don't know why type import loading order is different than regular ones. I personally use regular imports to import interfaces and other types so I don't have this issue.

**Comment by ljharb** [comment]
It’s a best practice to only ever use import type for type values, so anything we can do to make that easier, we should do.

**Comment by cherryblossom000** [comment]
I had a quick look at the code and it looks like type imports are ignored because of this line in `captureDependency`:

https://github.com/benmosher/eslint-plugin-import/blob/ff50964df1234515bc6f7969b90d87ada1982008/src/ExportMap.js#L414

This was introduced in #1494 so that `no-cycle` would ignore type imports.

Maybe a parameter `ignoreTypeImports` should be added to `ExportMap.parse` (and also `ExportMap.for` and `ExportMap.get`) so `no-cycle` can ignore type imports but `no-unused-modules` doesn't.

**Comment by ljharb** [comment]
@cherryblossom000 that sounds like a great idea. Want to make a PR?

**Comment by cherryblossom000** [comment]
@ljharb I'm working on it right now. However, I realised the solution I proposed above wouldn't work as the export maps are cached. I'm thinking of having a `typeOnly` property something like this:

```js
m.imports.set(p, {
  getter,
  source: { /* ... */ },
  importedSpecifiers,
  typeOnly: hasImportedType
}
```

**Comment by ljharb** [comment]
That also seems reasonable.

**Comment by cherryblossom000** [comment]
This is a bit more complicated than I expected. All these tests:

https://github.com/benmosher/eslint-plugin-import/blob/ff50964df1234515bc6f7969b90d87ada1982008/tests/src/rules/no-unused-modules.js#L806-L829

fail if this test is removed or moved to after the other tests:

https://github.com/benmosher/eslint-plugin-import/blob/ff50964df1234515bc6f7969b90d87ada1982008/tests/src/rules/no-unused-modules.js#L790-L805

In other words, the tests for files that are only imported as types only succeed when the file importing them is tested first.

I'm quite busy at the moment but I'll try to get around to why this happens as soon as I can.
